gstd/src/common/handlers.rs

```diff
@@ -25,50 +25,209 @@
+///     format!("panicked at '{MESSAGE}', {location}")
+/// }
+///
+/// #[rustversion::since(1.73)]
+/// fn expected_format(panic_info: &PanicInfo) -> String {
+///     let location = panic_info.location().unwrap();
+///     format!("panicked at {location}:\n{MESSAGE}")
+/// }
+///
+/// panic::set_hook(Box::new(|panic_info| {
+///     assert_eq!(panic_info.to_string(), expected_format(panic_info));
+/// }));
+///
+/// let result = panic::catch_unwind(|| {
+///     panic!("{MESSAGE}");
+/// });
+/// assert!(result.is_err());
+/// ```
+pub mod panic_handler {
+    #[cfg(target_arch = "wasm32")]
+    mod internal {
+        use crate::ext;
+        use core::panic::PanicInfo;
 
-    #[cfg(not(feature = "debug"))]
-    #[panic_handler]
-    pub fn panic(_: &PanicInfo) -> ! {
-        ext::panic("no info")
-    }
+        /// Panic handler when debug feature is disabled.
+        #[cfg(not(feature = "debug"))]
+        #[panic_handler]
+        pub fn panic(_: &PanicInfo) -> ! {
+            ext::panic("no info")
+        }
 
-    #[cfg(feature = "debug")]
-    #[panic_handler]
-    pub fn panic(panic_info: &PanicInfo) -> ! {
-        use crate::prelude::format;
-        #[cfg(not(feature = "panic-messages"))]
-        let message = None::<&core::fmt::Arguments<'_>>;
+        #[cfg(feature = "debug")]
+        mod constants {
+            /// Max amount of bytes allowed to be thrown as string explanation
+            /// of the error.
+            pub const TRIMMED_MAX_LEN: usize = 1024; //TODO: do not duplicate `gear_core::str::TRIMMED_MAX_LEN`
+            /// This prefix is used to print debug message:
+            /// `debug!("panic occurred: {msg}")`.
+            pub const PANIC_OCCURRED: &str = "panic occurred: ";
+            /// This prefix is used by `impl Display for PanicInfo<'_>`.
+            #[cfg(not(feature = "panic-messages"))]
+            pub const PANICKED_AT: &str = "panicked at ";
+        }
+
+        #[cfg(feature = "debug")]
+        use constants::*;
+
+        /// Panic handler for nightly Rust.
+        #[cfg(feature = "debug")]
         #[cfg(feature = "panic-messages")]
-        let message = panic_info.message();
-
-        let msg = match (message, panic_info.location()) {
-            (Some(msg), Some(loc)) => format!(
-                "'{:?}', {}:{}:{}",
-                msg,
-                loc.file(),
-                loc.line(),
-                loc.column()
-            ),
-            (Some(msg), None) => format!("'{msg:?}'"),
-            (None, Some(loc)) => {
-                format!("{}:{}:{}", loc.file(), loc.line(), loc.column())
+        #[panic_handler]
+        pub fn panic(panic_info: &PanicInfo) -> ! {
+            use crate::prelude::{fmt::Write, mem::MaybeUninit, str};
+            use arrayvec::ArrayString;
+
+            let option = panic_info.message().zip(panic_info.location());
+            let (message, location) = unsafe { option.unwrap_unchecked() };
```

Comment:
Don't use `unchecked` functions

Reply:
it always `Some`
https://github.com/rust-lang/rust/blob/5b8bc568d28b2e922290c9a966b3231d0ce9398b/library/core/src/panic/panic_info.rs#L26-L29
https://github.com/rust-lang/rust/blob/5b8bc568d28b2e922290c9a966b3231d0ce9398b/library/std/src/panicking.rs#L643-L644

Reply:
this has not changed for the last 5 years

Reply:
It doesn't matter `PanicInfo` always returns  `Some`. I'm talking about `unchecked` fns everywhere which can be considered UB in case of logic corruption

Reply:
I can add `SAFETY comment` to each unsafe block in that case

Reply:
9b2b9adf1263c4378f1bfe24de0b264f602e9846